### PR TITLE
feat(behavior_path_planner): pull over freespace parking

### PR DESCRIPTION
**Pull request by kosuke55**
Signed-off-by: kosuke55 <kosuke.tnp@gmail.com>

## Description

Add freespace parking to pull over module.
When the ego-vehicle is stuck with lane parking, execute feespace parking.

https://user-images.githubusercontent.com/39142679/218525084-f28b4dea-2bde-40b4-b229-4b745a5283b3.mp4

https://user-images.githubusercontent.com/39142679/218527647-061e9c8f-4d89-419c-9713-38c1ccf886e1.mp4

## Related links

https://github.com/tier4/autoware_launch/pull/771

## Tests performed

psim

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by TakaHoribe** [comment]
I'll check this

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 1]
```diff
@@ -90,11 +90,24 @@ PullOverModule::PullOverModule(
...
+    node.create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  lane_parking_timer_ = rclcpp::create_timer(
+    &node, clock_, lane_parking_period_ns, std::bind(&PullOverModule::onTimer, this),
+    lane_parking_timer_cb_group_);
+
+  // freespace parking
+  if (parameters_.enable_freespace_parking) {
+    freespace_planner_ = std::make_unique<FreespacePullOver>(node, parameters, vehicle_info_);
+    const auto freespace_parking_period_ns = rclcpp::Rate(1.0).period();
+    freespace_parking_timer_cb_group_ =
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:105]: What is the reason of making separate cb_groups with MutuallyExclusive?

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp [hunk 2]
```diff
@@ -44,17 +47,27 @@ using nav_msgs::msg::OccupancyGrid;
...
+  FREESPACE,
 };
 
 struct PUllOverStatus
 {
-  PullOverPath pull_over_path{};
+  std::shared_ptr<PullOverPath> pull_over_path{};
+  std::shared_ptr<PullOverPath> lane_parking_pull_over_path{};
   size_t current_path_idx{0};
+  bool require_increment_{true};
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp:70]: I couldn't understand the purpose of this flag. Would you add an explanation?

File: launch/tier4_planning_launch/launch/scenario_planning/parking.launch.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def launch_setup(context, *args, **kwargs):
                         "vehicle_frame": "base_link",
                         "map_frame": "map",
                         "update_rate": 10.0,
-                        "activate_by_scenario": True,
+                        "activate_by_scenario": False,
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply launch/tier4_planning_launch/launch/scenario_planning/parking.launch.py:65]: FYI: https://github.com/autowarefoundation/autoware.universe/pull/692

File: planning/behavior_path_planner/include/behavior_path_planner/path_utilities.hpp [hunk 4]
```diff
@@ -71,6 +72,26 @@ std::pair<TurnIndicatorsCommand, double> getPathTurnSignal(
   const ShiftLine & shift_line, const Pose & pose, const double & velocity,
   const BehaviorPathPlannerParameters & common_parameter);
 
+PathWithLaneId convertWayPointsToPathWithLaneId(
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply planning/behavior_path_planner/include/behavior_path_planner/path_utilities.hpp:75]: I will use this functions in pull out also.

File: planning/behavior_path_planner/README.md [hunk 5]
```diff
@@ -341,6 +341,14 @@ Generate two backward arc paths.
 | backward_parking_velocity                | [m/s] | double | velocity when backward parking                                            | -1.38         |
 | backward_parking_lane_departure_margin   | [m/s] | double | lane departure margin for front right corner of ego-vehicle when backward | 0.0           |
 
+###### Parameters freespace parking
+
+| Name                     | Unit | Type | Description                              | Default value |
+| :----------------------- | :--- | :--- | :--------------------------------------- | :------------ |
+| enable_freespace_parking | [-]  | bool | flag whether to enable freespace parking | true          |
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply planning/behavior_path_planner/README.md]: https://github.com/autowarefoundation/autoware.universe/pull/2879/commits/e5f3bea2732230fefefedf45a60f562d6301405d

File: planning/behavior_path_planner/src/path_utilities.cpp [hunk 6]
```diff
@@ -313,4 +313,157 @@ std::pair<TurnIndicatorsCommand, double> getPathTurnSignal(
   return std::make_pair(turn_signal, max_distance);
 }
 
+PathWithLaneId convertWayPointsToPathWithLaneId(
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply planning/behavior_path_planner/src/path_utilities.cpp:316]: no, these functions are for `PathWithLaneId` which is not used in `freespace_planner`, but these are similar to the ones of `freespace_planner`

File: planning/behavior_path_planner/src/path_utilities.cpp [hunk 7]
```diff
@@ -313,4 +313,157 @@ std::pair<TurnIndicatorsCommand, double> getPathTurnSignal(
...
+    const double dist = std::hypot(pose.position.x - p.x, pose.position.y - p.y);
+    if (dist < distance_threshold) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// only two points is supported
+std::vector<double> splineTwoPoints(
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply planning/behavior_path_planner/src/path_utilities.cpp:415]: hmm, I could not find it..

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp]: Added explanations and make the function in  https://github.com/autowarefoundation/autoware.universe/pull/2879/commits/3a74de4dd81529a074795f5c70a3a191d692fa58.

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:105]: [tier iv internal slack](https://star4.slack.com/archives/C03QW0GU6P7/p1677216635243619?thread_ts=1677212370.556509&cid=C03QW0GU6P7)

Both lane and freespace path generations take time to generate, so run them in separate threads.

File: planning/behavior_path_planner/README.md [hunk 8]
```diff
@@ -341,6 +341,14 @@ Generate two backward arc paths.
...
 | backward_parking_lane_departure_margin   | [m/s] | double | lane departure margin for front right corner of ego-vehicle when backward | 0.0           |
 
+###### Parameters freespace parking
+
+| Name                     | Unit | Type | Description                                                                                                          | Default value |
+| :----------------------- | :--- | :--- | :------------------------------------------------------------------------------------------------------------------- | :------------ |
+| enable_freespace_parking | [-]  | bool | This flag enables freespace parking, which runs when the vehicle is stuck due to e.g. obstacles in the parking area. | true          |
+
+See [reespace_planner](../freespace_planner/README.md) for other parameters.
+
```
Comment (kosuke55) [inline planning/behavior_path_planner/README.md:351]: docs is updated in https://github.com/autowarefoundation/autoware.universe/pull/2879
```suggestion
###### Parameters freespace parking

| Name                     | Unit | Type | Description                                                                                                          | Default value |
| :----------------------- | :--- | :--- | :------------------------------------------------------------------------------------------------------------------- | :------------ |
| enable_freespace_parking | [-]  | bool | This flag enables freespace parking, which runs when the vehicle is stuck due to e.g. obstacles in the parking area. | true          |

See [reespace_planner](../freespace_planner/README.md) for other parameters.

```

**Comment by kosuke55** [comment]
docs is updated in https://github.com/autowarefoundation/autoware.universe/pull/2879

File: planning/behavior_path_planner/include/behavior_path_planner/path_utilities.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply planning/behavior_path_planner/include/behavior_path_planner/path_utilities.hpp:75]: ok

File: launch/tier4_planning_launch/launch/scenario_planning/parking.launch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply launch/tier4_planning_launch/launch/scenario_planning/parking.launch.py:65]: thx

File: planning/behavior_path_planner/src/path_utilities.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (TakaHoribe) [reply planning/behavior_path_planner/src/path_utilities.cpp:415]: I misunderstood. Sorry for my bad.

**Review by TakaHoribe — APPROVED** [review]

**Pull request merged**